File: reference-python/src/Grocery_Sense/services/family_requests_service.py
"""
Grocery_Sense.services.family_requests_service

"Family picks": a household member (typically a kid / secondary) picks a meal
or an item; it lands on the shared shopping list attributed to them, and the
parent (master) gets a notification + review queue.

Design notes:
- No approval gate. Picks add to the list immediately; the parent reviews
  *after* (can remove). The unreviewed request count drives the parent badge.
- A request row is created only when the picker is a SECONDARY member, so the
  master adding their own things never self-notifies. Master picks still add to
  the list (so the feature is uniform), they just don't create a review item.
  The role decision lives here (single site) so the UI stays dumb.
- Hard excludes / allergies are household-wide and already enforced by
  preferences_service.get_meal_profile(); pickable_recipes() reuses that so a
  kid can never pick a recipe containing a household allergen.
"""

from __future__ import annotations

from typing import List, Optional

from Grocery_Sense.config import config_store
from Grocery_Sense.data.repositories import member_requests_repo
from Grocery_Sense.data.repositories.member_requests_repo import MemberRequestRow
from Grocery_Sense.recipes import recipe_engine
from Grocery_Sense.services import preferences_service
from Grocery_Sense.services.shopping_list_service import ShoppingListService

# add_single_item is a thin wrapper today; reuse it so a future auto_map upgrade
# flows through here too.
_shopping = ShoppingListService()


def _member_name(member_id: int) -> str:
    m = config_store.get_member(member_id)
    return m.name if m else f"Member {member_id}"


def _is_secondary(member_id: int) -> bool:
    m = config_store.get_member(member_id)
    return bool(m and m.role != config_store.ROLE_MASTER)


def pick_meal(member_id: int, recipe_name: str) -> Optional[MemberRequestRow]:
    """Add a recipe's ingredients to the shared list, attributed to the member.

    Returns the created request (secondary picker) or None (master picker).
    Raises ValueError if the recipe name is unknown — fail loud rather than add
    an empty pick.
    """
    recipe = recipe_engine.get_recipe_by_name(recipe_name)
    if recipe is None:
        raise ValueError(f"Unknown recipe: {recipe_name!r}")

    name = _member_name(member_id)
    ingredients = [str(i).strip() for i in (recipe.get("ingredients") or []) if str(i).strip()]

    row_ids: List[int] = []
    for ing in ingredients:
        row_ids.append(
            _shopping.add_single_item(
                name=ing,
                quantity=1.0,
                unit="each",
                notes=f"Family pick: {recipe_name}",
                added_by=name,
                added_by_member_id=member_id,
                auto_map=True,
            )
        )

    if not _is_secondary(member_id):
        return None

    req_id = member_requests_repo.add_request(
        member_id=member_id,
        member_name=name,
        kind="meal",
        label=recipe_name,
        item_row_ids=row_ids,
    )
    return member_requests_repo.get_request(req_id)


def pick_item(
    member_id: int,
    text: str,
    *,
    quantity: float = 1.0,
    unit: str = "each",
) -> Optional[MemberRequestRow]:
    """Add a single item to the shared list, attributed to the member.

    Returns the created request (secondary picker) or None (master picker).
    """
    label = (text or "").strip()
    if not label:
        raise ValueError("Item text is required.")

    name = _member_name(member_id)
    row_id = _shopping.add_single_item(
        name=label,
        quantity=quantity,
        unit=unit or "each",
        notes="Family pick",
        added_by=name,
        added_by_member_id=member_id,
        auto_map=True,
    )

    if not _is_secondary(member_id):
        return None

    req_id = member_requests_repo.add_request(
        member_id=member_id,
        member_name=name,
        kind="item",
        label=label,
        item_row_ids=[row_id],
    )
    return member_requests_repo.get_request(req_id)


def pickable_recipes() -> List[str]:
    """Recipe names a member may pick: household hard excludes / allergies hidden.

    Household-wide (allergies block for everyone), so no per-member arg. Soft
    excludes do NOT filter — a kid can still pick those.
    """
    recipes = recipe_engine.load_all_recipes()
    if not recipes:
        return []

    all_ingredients = {
        str(i).strip()
        for r in recipes
        for i in (r.get("ingredients") or [])
        if str(i).strip()
    }
    profile = preferences_service.get_meal_profile()
    # include_ingredients = every ingredient → every recipe overlaps, so this
    # returns all recipes that survive the hard-profile filter.
    passing = recipe_engine.filter_recipes_by_ingredients_and_profile(
        include_ingredients=all_ingredients,
        profile=profile,
        max_results=len(recipes),
    )
    names = [str(r.get("name", "")).strip() for r in passing if str(r.get("name", "")).strip()]
    return sorted(names, key=str.lower)


# --- thin pass-throughs for the UI ----------------------------------------

def unreviewed_count() -> int:
    return member_requests_repo.count_unreviewed()


def list_unreviewed() -> List[MemberRequestRow]:
    return member_requests_repo.list_unreviewed()


def mark_reviewed(request_id: int) -> None:
    member_requests_repo.mark_reviewed(request_id)


def remove_request(request_id: int) -> None:
    """Soft-delete the shopping_list rows this pick created, then mark reviewed."""
    req = member_requests_repo.get_request(request_id)
    if req is None:
        return
    for row_id in req.item_row_ids:
        _shopping.soft_delete_item(row_id)
    member_requests_repo.mark_reviewed(request_id)

File: Grocery_Sense/GrocerySense.Core/Services/FamilyRequestsService.cs
using GrocerySense.Data;
using GrocerySense.Data.Repositories;
using GrocerySense.Domain;
using Microsoft.Data.Sqlite;

namespace GrocerySense.Core;

// Port of reference-python/.../services/family_requests_service.py — "family picks": a household member
// picks a meal or item, it lands on the shared shopping list attributed to them, and the parent (master)
// reviews after the fact. No approval gate — picks add immediately; a review row is created ONLY for a
// SECONDARY picker (the master adding their own things never self-notifies). Household hard excludes /
// allergies are enforced by PickableRecipes so a kid can't pick an allergen recipe.
public sealed class FamilyRequestsService
{
    private readonly ConfigStore _config;
    private readonly RecipeEngine _engine;
    private readonly PreferencesService _preferences;
    private readonly IngredientMappingService _mapper;
    private readonly SqliteConnectionFactory _factory;
    private readonly MealSuggestionService? _meals; // null => PickableRecipesRanked falls back alphabetical

    public FamilyRequestsService(ConfigStore config, RecipeEngine engine,
        PreferencesService preferences, IngredientMappingService mapper, SqliteConnectionFactory factory,
        MealSuggestionService? meals = null)
    {
        _config = config;
        _engine = engine;
        _preferences = preferences;
        _mapper = mapper;
        _factory = factory;
        _meals = meals;
    }

    // Add a recipe's ingredients to the shared list, attributed to the member. Returns the created request
    // (secondary picker) or null (master picker). Throws on an unknown recipe — fail loud, not an empty pick.
    // Ingredient rows + the review request commit in ONE transaction: a mid-pick failure must not leave a
    // partial meal on the list (or orphaned rows with no request to undo them from).
    public MemberRequestRow? PickMeal(int memberId, string recipeName)
    {
        var recipe = _engine.GetRecipeByName(recipeName)
            ?? throw new ArgumentException($"Unknown recipe: {recipeName}");

        // Defense-in-depth: re-check the household hard filter at pick time. PickableRecipes filters the list,
        // but a stale list (e.g. an allergy added while a kid had /family open) could still surface a now-unsafe
        // recipe — refuse rather than add an allergen to the shared list. Mirrors MealSuggestionService's net.
        if (ProfileFilter.Violates(recipe.Ingredients, _preferences.GetMealProfile()))
            throw new InvalidOperationException(
                $"\"{recipe.Name}\" is no longer allowed by the household's allergy/exclude settings.");

        var name = MemberName(memberId);
        // Map before the write transaction (match-only; the mapper opens its own connection).
        var mapped = recipe.Ingredients.Select(ing => (Name: ing, ItemId: _mapper.MapToItem(ing).ItemId)).ToList();

        MemberRequestRow? request;
        using (var conn = _factory.Open())
        using (var tx = conn.BeginTransaction())
        {
            var rowIds = mapped
                .Select(m => ShoppingListRepo.AddItem(conn, m.Name, 1.0, "each", notes: $"Family pick: {recipeName}",
                    addedBy: name, addedByMemberId: memberId, itemId: m.ItemId, tx: tx))
                .ToList();
            request = CreateRequestIfSecondary(conn, tx, memberId, name, "meal", recipeName, rowIds);
            tx.Commit();
        }
        _mapper.FlushLearnedAliases(); // after commit — the flush opens its own write connection
        return request;
    }

    // Add a single item to the shared list, attributed to the member.
    public MemberRequestRow? PickItem(int memberId, string text, double quantity = 1.0, string unit = "each")
    {
        var label = (text ?? "").Trim();
        if (label.Length == 0) throw new ArgumentException("Item text is required.");

        var name = MemberName(memberId);
        var itemId = _mapper.MapToItem(label).ItemId;

        MemberRequestRow? request;
        using (var conn = _factory.Open())
        using (var tx = conn.BeginTransaction())
        {
            var rowId = ShoppingListRepo.AddItem(conn, label, quantity, string.IsNullOrEmpty(unit) ? "each" : unit,
                notes: "Family pick", addedBy: name, addedByMemberId: memberId, itemId: itemId, tx: tx);
            request = CreateRequestIfSecondary(conn, tx, memberId, name, "item", label, new[] { rowId });
            tx.Commit();
        }
        _mapper.FlushLearnedAliases();
        return request;
    }

    // Recipe names a member may pick: household hard excludes / allergies are hidden (whole-household, so no
    // per-member arg). Soft excludes do NOT filter. Sorted case-insensitively.
    public IReadOnlyList<string> PickableRecipes() =>
        _engine.RecipesMatchingProfile(_preferences.GetMealProfile())
            .Select(r => r.Name).Where(n => n.Length > 0)
            .OrderBy(n => n, StringComparer.OrdinalIgnoreCase).ToList();

    // Pickable recipes ranked by the meal-suggestion value score, so kids choose from what's cheap
    // this week. Both lists filter through the same ProfileFilter, so every pickable name gets a score.
    // No MealSuggestionService (older tests) -> alphabetical, unflagged — same behavior as before.
    public IReadOnlyList<PickableRecipe> PickableRecipesRanked()
    {
        var names = PickableRecipes();
        if (_meals is null || names.Count == 0)
            return names.Select(n => new PickableRecipe(n, false)).ToList();

        var scored = _meals.SuggestMealsForWeek(_preferences.GetMealProfile(), maxRecipes: int.MaxValue);
        var byName = scored.ToDictionary(s => s.Recipe.Name, StringComparer.OrdinalIgnoreCase);
        return names
            .OrderByDescending(n => byName.TryGetValue(n, out var s) ? s.TotalScore : double.MinValue)
            .Select(n => new PickableRecipe(n,
                byName.TryGetValue(n, out var s) && s.DealScore > 0.2)) // 0.2 = FormatMealExplanation's on-sale bar
            .ToList();
    }

    // --- parent review queue ---

    public int UnreviewedCount()
    {
        using var conn = _factory.Open();
        return MemberRequestsRepo.CountUnreviewed(conn);
    }

    public IReadOnlyList<MemberRequestRow> ListUnreviewed()
    {
        using var conn = _factory.Open();
        return MemberRequestsRepo.ListUnreviewed(conn);
    }

    public void MarkReviewed(int requestId)
    {
        using var conn = _factory.Open();
        MemberRequestsRepo.MarkReviewed(conn, requestId);
    }

    // Undo a pick: soft-delete the shopping_list rows it created + mark reviewed, in one transaction.
    public void RemoveRequest(int requestId)
    {
        using var conn = _factory.Open();
        var req = MemberRequestsRepo.GetRequest(conn, requestId);
        if (req is null) return;
        using var tx = conn.BeginTransaction();
        foreach (var rowId in req.ItemRowIds) ShoppingListRepo.DeleteItem(conn, rowId, tx);
        MemberRequestsRepo.MarkReviewed(conn, requestId, tx);
        tx.Commit();
    }

    private MemberRequestRow? CreateRequestIfSecondary(SqliteConnection conn, SqliteTransaction tx,
        int memberId, string name, string kind, string label, IReadOnlyList<int> rowIds)
    {
        if (!_config.IsSecondary(memberId)) return null; // master picks don't create a review row
        var reqId = MemberRequestsRepo.AddRequest(conn, memberId, name, kind, label, rowIds, tx);
        return MemberRequestsRepo.GetRequest(conn, reqId, tx);
    }

    private string MemberName(int memberId) => _config.GetMember(memberId)?.Name is { Length: > 0 } n
        ? n : $"Member {memberId}";
}
